CMakeLists.txt

```diff
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: BSD-3-Clause
 
-cmake_minimum_required(VERSION 3.10)
+cmake_minimum_required(VERSION 3.20.0)
```

Comment:
This is probably going to cause some CI failures.

Reply:
I had submitted a request to SOF infrastructure team for cmake upgrade.

Reply:
@marc-hb @greg-intel I assume Jenkins is good for this cmake version ?

Reply:
It seems to be failing not just in Jenkins: https://github.com/thesofproject/sof/pull/5265

Reply:
Can you require 3.20 for `CMAKE_HOST_WIN32` only? According to your commit message it does not seem needed for Linux.

Reply:
@marc-hb we are moving to Zephyr being the default rtos. We need to plan and upgrade the containers so that all developers can build for Zephyr. 
Let also assume that container updates will be ad-hoc and follow upstream development. This should be factored into our processes so it's a painless as possible.

Reply:
We offer only one crosstool-ng container for developers, so this single one is not the problem. The problem is the large variety of Linux build environments (not just containers) across XTOS developers and XTOS build machines in multiple CIs maintained by different owners, some we don't even know about. Breaking the build for _all_ these legacy XTOS environments  unless they upgrade to cutting edge CMake version is in aggregate a huge disruption and amount of work for... just one Windows-only command? Really not worth it when this command can be guarded and used by Windows only. Seems like a no-brainer.

Zephyr requested a cutting edge CMake version because it really needed some critical CMake features yet users were not happy. Unlike XTOS, Zephyr developers got at least a very active stream of new features in return.

Off-topic but note many developers do not use our crosstool-ng container at all, because either
- they use xcc, or
- they built crosstool-ng themselves to avoid the docker layer of indirection and immutability of the container (very painful to install extra packages or any other image change), or
- they use Zephyr which does not compile with our crosstool-ng.

BTW the Zephyr SDK is an open-source project also based on crosstool-ng, so why are still maintaining our own crosstool-ng fork? (and container?) Or should we at least rebase our crosstool-ng on top of the Zephyr SDK? Off-topic sorry.

Reply:
> @marc-hb @greg-intel I assume Jenkins is good for this cmake version ?

Five of the build servers, sofbld07.ostc.intel.com through sofbld11.ostc.intel.com are running cmake version 3.22.1, but the others are using 3.16.3, so it's hit or miss.

Reply:
> Breaking the build for all these legacy XTOS environments unless they upgrade to cutting edge CMake version is in aggregate a huge disruption and amount of work for... just one Windows-only command? Really not worth it when this command can be guarded and used by Windows only. Seems like a no-brainer.

Windows command is not a cause of this change but a reason to do it. So is Zephyr. I believe that this is what software maintenance is about - we constantly try to update all our tools not to get obsolete. Same with python, testing frameworks etc. It is good to perform this step now before further build system refactorings so we could check all our CI supported configurations are passing after the upgrade. Besides as @greg-intel mentioned, some servers already run CMake 3.22.1.

Reply:
> So is Zephyr.

Zephyr is totally off-topic here as it already requires 3.21 (which didn't please a number of users). This particular discussion is about XTOS developers and XTOS only.

> Windows command is not a cause of this change but a reason to do it. 

So move cmake_minimum_required(3.20) to the `if(CMAKE_HOST_WIN32)` section for now and revisit and annoy ALL Linux users (giving Intel a bad name) later when you have much stronger reason(s) to do so. If you wait long enough then 3.20 will be available in most Linux distributions by default and this will have become a non-issue. How inconvenient is this to you and how does that inconvenience compares to the inconvenience of every single XTOS developer, virtual machine and CI system inside and outside Intel to add one extra repository and upgrade one of their most fundamental development packages, potentially affecting all their other development efforts?

> I believe that this is what software maintenance is about - we constantly try to update all our tools not to get obsolete. 

This is clearly the point of view of a developer: someone who enjoys writing and rewriting code and can afford to spend most of his time doing that. People swamped with applying security fixes and finding and reporting regressions have a very different perspective.

This vision of software maintenance rings truer to me:
https://medium.com/@sdboyer/so-you-want-to-write-a-package-manager-4ae9c17d9527

> Despite the lack of necessary relationship between versions and code, there is at least one way in which versions help quite directly to ensure software works: they focus public attention on a few particular revisions of code. With attention focused thusly, Linus’ Law suggests that bugs will be rooted out and fixed (and released in subsequent patch versions). In this way, the curatorial effect of focusing attention on particular versions reduces systemic risk around those versions. This helps with another of FLOSS’ uncertainties:

> > I have to prepare myself for the likelihood that most of what’s out there will probably be crap. Sorting wheat from chaff will also take time.

> Having versions at least ensures that, if the software is crap, it’s because it’s actually crap, not because you grabbed a random revision that happened to be crap. **That saves time. Saved time can save projects.**

Reply:
This discussion is going nowhere.
Thankfully we do not have to agree and you are not decisive person here.
> So move cmake_minimum_required(3.20) to the if(CMAKE_HOST_WIN32) section for now and revisit and annoy ALL Linux users (giving Intel a bad name) later when you have much stronger reason(s) to do so.

Look at Intel slogans. I see there "Fearless". I see there "newest technologies".
Do you think your approach towards cmake aligns with this policy and culture?
We need to go forward not lag behind. It's not our problem that linux package managers lags behind.
What kind of approach it is? You install software - always newest stable version. Thats trivial and obvious for everyone - for Windows users, for Mac users, for Android users but no, not for linux package managers.

Reply:
> Do you think your approach [...] ? It's not our problem that linux package managers lags behind.

I'm absolutely not claiming Linux package management is the best approach[*], please stop identify me with that. Don't confuse me with Ubuntu, I'm just the messenger. I'm merely describing where all existing SOF users are, inside and outside Intel and I think they matter. In fact many Linux users are not happy either: https://lwn.net/Articles/867657/

Package management is always very messy no matter how you do it and Windows is no exception. One of its (universal) aspect is called "DLL hell". It's not a solved problem and will never be. There are pros and cons in all approaches. I highly recommend Sam Boyer's article I referenced above if you're interested in a broader perspective and more informed and nuanced opinions. 

[*] there are "rolling" distros too like Arch, Gentoo and others but they are much less popular.

> Thats obvious for Windows users,...

Microsoft is very actively supporting WSL2 and the default choice is Ubuntu; I find that very interesting. BTW does Visual Studio finally support C99?
> Look at Intel slogans. I see there "Fearless". I see there "newest technologies".

Intel also has a generally good standing in open-source communities outside Intel and "Wintel"'s backward compatibility has no equivalent in the entire industry.

This topic is one of the most complicated.